File: app/src/main/java/com/kooritea/fcmfix/xposed/KeepNotification.java
package com.kooritea.fcmfix.xposed;

import android.os.Build;
import android.service.notification.NotificationListenerService;

import java.lang.reflect.Method;
import com.kooritea.fcmfix.libxposed.XC_MethodHook;
import com.kooritea.fcmfix.libxposed.XposedBridge;
import com.kooritea.fcmfix.libxposed.XposedHelpers;

/**
 * 通知保持模块 - 防止系统自动清除 FCM 通知
 * 
 * 功能说明：
 * 部分 Android 系统会在应用未启动或后台运行时自动清除其通知，导致用户无法看到 FCM 推送。
 * 本模块通过 Hook 系统的通知管理服务，阻止系统自动清除目标应用的通知。
 * 
 * 工作原理：
 * 1. 查找 NotificationManagerService 类中的 cancelAllNotificationsInt 方法
 * 2. 根据不同 Android 版本确定参数位置（包名和原因代码）
 * 3. 当检测到目标应用且原因是 PACKAGE_CHANGED 或特定代码时，阻止取消通知
 * 
 * 支持的 Android 版本：
 * - Android 10 (API 30) 到 Android 15 (API 35) 及以上
 * - 自动适配不同版本的方法签名变化
 * 
 * 配置项：
 * - disableAutoCleanNotification: 启用后阻止系统自动清除通知
 * 
 * @author fcmfix
 */
public class KeepNotification extends XposedModule{

    public KeepNotification(ClassLoader classLoader) {
        super(classLoader);
        try {
            this.startHook();
        } catch (Throwable e) {
            printLog("No Such Method com.android.server.notification.NotificationManagerService.cancelAllNotificationsInt");
        }
    }
    
    /**
     * 开始 Hook 操作
     * Hook 通知取消方法，防止系统自动清除目标应用的通知
     * 
     * @throws NoSuchMethodError 当找不到目标方法时抛出
     * @throws XposedHelpers.ClassNotFoundError 当找不到目标类时抛出
     */
    protected void startHook() throws NoSuchMethodError, XposedHelpers.ClassNotFoundError {
        // 查找通知管理服务类
        Class<?> clazz = XposedHelpers.findClass("com.android.server.notification.NotificationManagerService",classLoader);
        final Method[] declareMethods = clazz.getDeclaredMethods();
        Method targetMethod = null;
        
        // 查找 cancelAllNotificationsInt 方法，选择参数最多的版本（通常是最新的）
        for(Method method : declareMethods){
            if("cancelAllNotificationsInt".equals(method.getName())){
                if(targetMethod == null || targetMethod.getParameterTypes().length < method.getParameterTypes().length){
                    targetMethod = method;
                }
            }
        }
        
        if(targetMethod != null){
            int pkg_args_index = 0;  // 包名参数索引
            int reason_args_index = 0;  // 原因代码参数索引
            
            // 根据不同 Android 版本设置参数索引
            if(Build.VERSION.SDK_INT == 30){
                pkg_args_index = 2;
                reason_args_index = 8;
            }
            if(Build.VERSION.SDK_INT == 31){
                pkg_args_index = 2;
                reason_args_index = 8;
            }
            if(Build.VERSION.SDK_INT == 32){
                pkg_args_index = 2;
                reason_args_index = 8;
            }
            if(Build.VERSION.SDK_INT == 33){
                pkg_args_index = 2;
                reason_args_index = 8;
            }
            if(Build.VERSION.SDK_INT == 34){
                if(targetMethod.getParameterTypes().length == 10){
                    pkg_args_index = 2;
                    reason_args_index = 8;
                }else if(targetMethod.getParameterTypes().length == 8){
                    pkg_args_index = 2;
                    reason_args_index = 7;
                }
            }
            if(Build.VERSION.SDK_INT == 35){
                pkg_args_index = 2;
                reason_args_index = 7;
            }
            if(Build.VERSION.SDK_INT > 35){
                pkg_args_index = 2;
                reason_args_index = 7;
            }
            
            // 如果参数索引未正确设置，抛出异常
            if(pkg_args_index == 0 || reason_args_index == 0){
                throw new NoSuchMethodError();
            }
            
            int finalPkg_args_index = pkg_args_index;
            int finalReason_args_index = reason_args_index;
            
            // Hook 目标方法
            XposedBridge.hookMethod(targetMethod,new XC_MethodHook() {
                @Override
                protected void beforeHookedMethod(MethodHookParam param) {
                    // 检查系统是否已完成启动
                    if(!isBootComplete){
                        return;
                    }
                    
                    // 如果启用了禁用自动清理通知功能且是目标应用
                    if(getBooleanConfig("disableAutoCleanNotification",false) && targetIsAllow((String) param.args[finalPkg_args_index])){
                        int reason = (int)param.args[finalReason_args_index];
                        
                        // 如果原因是应用包变化（如更新），阻止取消通知
                        if(reason == NotificationListenerService.REASON_PACKAGE_CHANGED){
                            param.setResult(null);
                        }
                        
                        // 处理 ColorOS 15 / OxygenOS 15 的特定原因代码
                        if(reason == 10020 || reason == 10021){ // cos15/oos15
                            param.setResult(null);
                        }
                    }
                }
            });
        }else{
            throw new NoSuchMethodError();
        }
    }
}

File: app/src/main/kotlin/sumicya/fcmself/mods/KeepNotification.kt
package sumicya.fcmself.mods

import android.os.Build
import android.service.notification.NotificationListenerService

import io.github.libxposed.api.XposedInterface

import sumicya.fcmself.FcmselfModule
import sumicya.fcmself.ProcessEnv
import sumicya.fcmself.core.FcmselfLog
import sumicya.fcmself.core.Push
import sumicya.fcmself.hook.Hooks
import sumicya.fcmself.hook.Reflect
import sumicya.fcmself.hook.Signatures

/**
 * KeepNotification —— 防止系统自动清除推送通知。
 *
 * 部分 ROM 会在应用未运行时因「包变化」（更新/卸载等）自动清掉其通知。原生 AOSP
 * 没有这个行为，本模块 Hook `NotificationManagerService.cancelAllNotificationsInt`，
 * 在取消原因为 REASON_PACKAGE_CHANGED（或 ColorOS / OxygenOS 特有的 10020 / 10021）
 * 时直接忽略这次取消，把行为还原成原生语义——对所有包生效（无白名单）。
 *
 * (pkg, reason) 下标由 [Signatures.resolveNotificationArgs] 自适应解析：
 * 候选顺序按 API 版本给出（34 的 10 参形态 reason@8 与 8/9 参形态 reason@7 两代），
 * 全部候选过类型校验，不符则安全跳过并打出完整签名。
 */
internal class KeepNotification(api: XposedInterface, classLoader: ClassLoader) :
    FcmselfModule(api, classLoader, "KeepNotification") {

    init {
        point("NotificationManagerService#cancelAllNotificationsInt") { startHook() }
    }

    private fun startHook() {
        val className = "com.android.server.notification.NotificationManagerService"
        val clazz = Reflect.findClass(className, classLoader)
        val method = Reflect.findMethodMostParams(clazz, "cancelAllNotificationsInt")
            ?: throw NoSuchMethodError("$className#cancelAllNotificationsInt")

        // reason 候选顺序承载消歧规则：<=33 固定 @8；34 先试 @8（10 参）再试 @7（8 参）；
        // 35+ 先试 @7（9 参）。每个候选都会经 MethodArgs 校验，全部失败则放弃挂载。
        val reasonCandidates = when {
            Build.VERSION.SDK_INT <= 33 -> listOf(8)
            Build.VERSION.SDK_INT == 34 -> listOf(8, 7)
            else -> listOf(7, 8)
        }
        val (pkgIndex, reasonIndex) = Signatures.resolveNotificationArgs(method, reasonCandidates) {
            FcmselfLog.log(it)
        } ?: return
        FcmselfLog.log(
            "cancelAllNotificationsInt hook 参数：pkg@$pkgIndex reason@$reasonIndex（API ${Build.VERSION.SDK_INT}）"
        )

        Hooks.hook(api, method) { chain ->
            // 系统启动完成前不介入
            if (!ProcessEnv.isBootComplete) return@hook chain.proceed()
            val pkg = chain.getArg(pkgIndex) as? String
            if (!Push.hasTarget(pkg)) return@hook chain.proceed()
            val reason = chain.getArg(reasonIndex) as? Int
            // 原因是应用包变化（更新/卸载）：阻止取消，等于「这次取消请求被忽略」
            if (reason == NotificationListenerService.REASON_PACKAGE_CHANGED ||
                reason == REASON_COS_OOS_1 ||
                reason == REASON_COS_OOS_2
            ) {
                return@hook null
            }
            chain.proceed()
        }
    }

    private companion object {
        /** ColorOS 15 / OxygenOS 15 特有的取消原因代码。 */
        const val REASON_COS_OOS_1 = 10020
        const val REASON_COS_OOS_2 = 10021
    }
}
